core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -3454,15 +3454,23 @@ object ZIO extends ZIOCompanionPlatformSpecific with ZIOCompanionVersionSpecific
   final def foreach[R, E, A, B: ClassTag](in: Array[A])(f: A => ZIO[R, E, B])(implicit
     trace: Trace
   ): ZIO[R, E, Array[B]] =
-    if (in.isEmpty) ZIO.succeed(Array.empty[B])
-    else
-      ZIO.suspendSucceed {
-        val iterator = in.iterator
-        val builder  = Array.newBuilder[B]
-        builder.sizeHint(in.length)
-
-        ZIO.whileLoop(iterator.hasNext)(f(iterator.next()))(builder += _).as(builder.result())
-      }
+    in.length match {
+      case 0 => ZIO.succeed(Array.empty[B])
+      case 1 => ZIO.suspendSucceed(f(in(0)).map(Array(_)))
+      case size =>
+        ZIO.suspendSucceed {
+          val iterator = in.iterator
+          val arr      = Array.ofDim[B](size)
+          var i        = 0
+
+          ZIO
+            .whileLoop(iterator.hasNext)(f(iterator.next())) { b =>
+              arr(i) = b
+              i += 1
+            }
+            .as(arr)
```

Comment:
Do we have an `as` operator that doesn't suspend the value?

```suggestion
            .flatMap(_ => Exit.succeed(arr))
```

Reply:
No, I think your suggestion is the best that we can do

Reply:
Although it wouldn't work without some sort of suspension as `arr` would otherwise be read before the whileLoop is evaluated

Reply:
On second thought, I'm going to keep this as is for now and revisit this in a separate PR as there are multiple places that this is occuring